src/integrationTest/codewhisperer/serviceInvocations.test.ts

```diff
@@ -0,0 +1,133 @@
+/*!
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as assert from 'assert'
+import * as vscode from 'vscode'
+import { setValidConnection } from '../util/codewhispererUtil'
+import { ConfigurationEntry } from '../../codewhisperer/models/model'
+import * as codewhispererClient from '../../codewhisperer/client/codewhisperer'
+import { RecommendationHandler } from '../../codewhisperer/service/recommendationHandler'
+import {
+    createMockTextEditor,
+    createTextDocumentChangeEvent,
+    resetCodeWhispererGlobalVariables,
+} from '../../test/codewhisperer/testUtil'
+import { KeyStrokeHandler } from '../../codewhisperer/service/keyStrokeHandler'
+import { sleep } from '../../shared/utilities/timeoutUtils'
+import { invokeRecommendation } from '../../codewhisperer/commands/invokeRecommendation'
+
+describe('CodeWhisperer service invocation', async function () {
+    let validConnection: boolean
+    const client = new codewhispererClient.DefaultCodeWhispererClient()
+    const config: ConfigurationEntry = {
+        isShowMethodsEnabled: true,
+        isManualTriggerEnabled: true,
+        isAutomatedTriggerEnabled: true,
+        isSuggestionsWithCodeReferencesEnabled: true,
+    }
+
+    before(async function () {
+        //valid connection required to run tests
+        validConnection = await setValidConnection()
+    })
+
+    beforeEach(function () {
+        resetCodeWhispererGlobalVariables()
+        RecommendationHandler.instance.clearRecommendations()
+    })
```

Comment:
This should work without having to duplicate the condition across every test:
```suggestion
    beforeEach(function () {
        resetCodeWhispererGlobalVariables()
        RecommendationHandler.instance.clearRecommendations()
        if (!validConnection && this.currentTest) {
            this.currentTest.title += ` (skipped - no valid connection)`
            this.currentTest.skip()
        }
    })
```
You can make this into a util function too:
```ts
async function setConnectionOrSkipTest(ctx: Mocha.Context) {
    const conn = await setValidConnection()
    if (!conn && this.currentTest) {
       ctx.currentTest.title += ` (skipped - no valid connection)`
       ctx.currentTest.skip()
    }
}

// In `beforeEach`
validConnection = await setConnectionOrSkipTest(this)
```

Reply:
Thanks for this input. I added this logic to the util file.

I changed the function slightly so that the valid connection is still set in a different function, because I found that if I make the `beforeEach` async and include the `skip` there I get the error `sync skip; aborting execution`.

so the updated function is:

```
export function skiptTestIfNoValidConn(validConnection: boolean, ctx: Mocha.Context) {
    if (!validConnection && ctx.currentTest) {
        ctx.currentTest.title += ` (skipped - no valid connection)`
        ctx.currentTest.skip()
    }
}
```
invocation: 

```
before(async function () {
        validConnection = await setValidConnection()
    })

beforeEach(function () {
        ...
        skiptTestIfNoValidConn(validConnection, this)
    })
```